### PR TITLE
Don't skip elaborations when reporting errors for cached failed relations

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/3276
to learn more about what's happening you can read my comment [here](https://github.com/microsoft/TypeScript/issues/3276#issuecomment-1661197136)

I'm just opening this as a draft, even without updating baselines. It might be interesting to run the perf suite on this one.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21205,7 +21205,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const id = getRelationKey(source, target, intersectionState, relation, /*ignoreConstraints*/ false);
             const entry = relation.get(id);
             if (entry !== undefined) {
-                if (reportErrors && entry & RelationComparisonResult.Failed && !(entry & RelationComparisonResult.Reported)) {
```
Comment (Andarist) [inline src/compiler/checker.ts:21208]: if this fix turns out to be in a good direction then `RelationComparisonResult.Reported` should be cleaned up further

**Comment by RyanCavanaugh** [comment]
@typescript-bot perf test this faster

**Comment by Andarist** [comment]
@RyanCavanaugh since the perf run came up clean I updated the baselines and cleaned up this a little bit further. Looking at the changed baselines it seems like a neat improvement to me :)

**Review by weswigham — APPROVED** [review]

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot test top100
@typescript-bot new perf test this

**Comment by sandersn** [comment]
@weswigham @ahejlsberg I'd like to merge this at the start of 5.4 development. Is it good to go?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -20933,7 +20931,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (overflow) {
             // Record this relation as having failed such that we don't attempt the overflowing operation again.
             const id = getRelationKey(source, target, /*intersectionState*/ IntersectionState.None, relation, /*ignoreConstraints*/ false);
-            relation.set(id, RelationComparisonResult.Reported | RelationComparisonResult.Failed);
+            relation.set(id, RelationComparisonResult.Failed);
```
[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts:20934]: Hmm, the CI shows that this actually is important as now it's doing a lot of heavy work twice in this pathological case and the CI timeouts.

If we take a closer look at this baseline (even on `main`) we might notice that it's actually reporting 2 errors at the same location:
https://github.com/microsoft/TypeScript/blob/7216ee0bb821b4b565d54c52dcbf788a3fdf7062/tests/baselines/reference/relationComplexityError.errors.txt#L1-L2

So perhaps If I manage to report this once, the perf problems will go away naturally.

**Comment by sandersn** [comment]
@typescript-bot perf test this

**Comment by gabritto** [comment]
We had this PR all this time and never merged it? Oh well.

**Comment by Andarist** [comment]
@gabritto does it overlap with some other fresh work/issue? I can certainly try to prioritize this PR more and bring it up to speed

**Comment by jakebailey** [comment]
Yeah, #57842, since we may report diagnostics out of order and therefore elaborations may jump around.

**Comment by gabritto** [comment]
> Yeah, #57842, since we may report diagnostics out of order and therefore elaborations may jump around.

To give more details, I'm doing some diagnostics work in preparation for #57842. Right now I'm trying to have less duplicated diagnostics, but next on my list is to provide assignability elaboration consistently in editor scenarios, and as a consequence get rid of some of the diagnostics differences that arise when using #57842 vs not using them.
Where things might overlap is that the fact that assignability diagnostics varies depending on cached relations sometimes makes it harder to detect duplication, so making the assignability diagnostics more consistent could help with that as well.

**Comment by Andarist** [comment]
This work has landed as part of https://github.com/microsoft/TypeScript/pull/58859

**Comment by gabritto** [comment]
> This work has landed as part of #58859

Thanks for doing this!

**Pull request closed without merging**